### PR TITLE
Change elmo embedding to use tfhub

### PR DESCRIPTION
**Pull request by dorukkilitcioglu**
This PR fixes the ELMo issues by changing the source of ELMo embeddings to TFHub from AllenNLP. This means one less dependency.

It also includes a quick fix for a new sklearn-related issue, wherein it doesn't accept `np.matrix` inputs anymore.

### CODE REVIEW

**Comment by dorukkilitcioglu** [comment]
Quick holiday PR for all @skadio @kuppulur. This should fix the issue with ELMo embeddings and make the full installation work.

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -15,7 +15,7 @@ jobs:
     runs-on: ubuntu-latest
     strategy:
       matrix:
-        python-version: [3.6, 3.7, 3.8, 3.9]
+        python-version: ['3.7', '3.8', '3.9', '3.10']
```
Comment (skadio) [inline .github/workflows/ci.yml:18]: let's make sure we note Python 3.7+ in right places (readme/docs etc.)

File: CHANGELOG.txt [hunk 2]
```diff
@@ -2,6 +2,12 @@
 TextWiser CHANGELOG
 =====================
 
+-------------------------------------------------------------------------------
+Dec 19, 2022 1.4.1
+-------------------------------------------------------------------------------
+minor:
```
Comment (skadio) [inline CHANGELOG.txt]: change of backend seems major, to me

File: requirements_full.txt [hunk 3]
```diff
@@ -1,5 +1,3 @@
-allennlp>=1.0,<2.7.0; python_version == '3.6'
```
Comment (skadio) [inline requirements_full.txt:1]: nice!

File: textwiser/_version.py [hunk 4]
```diff
@@ -1,4 +1,4 @@
 # Copyright 2019 FMR LLC <opensource@fidelity.com>
 # SPDX-License-Identifer: Apache-2.0
 
-__version__ = "1.4.0"
+__version__ = "1.4.1"
```
Comment (skadio) [inline textwiser/_version.py]: same comment on minor/major change

**Review by skadio — APPROVED** [review]
LGTM. Thanks for the quick fix. 

High-level comment; quite happy to see how relatively lean it is, to replace one embedding within the library.

**Comment by skadio** [comment]
Unrelated comment; it seems we have an Open Issue, can we plz close it accordingly

File: .github/workflows/ci.yml (hunk 1 above)
Comment (dorukkilitcioglu) [reply .github/workflows/ci.yml:18]: I don't think dropping the 3.6 testing _necessarily_ means that we say the library is 3.7+. It still works on 3.6 - we just can't test it automatically. Would you rather be on the safer side and drop it altogether?

Comment (skadio) [reply .github/workflows/ci.yml:18]: oh I see.. I got confused with Seq2Pat, as in that case, 3.6 does not work, hence dropped fully jumping to 3.7 in the docs/readme/requirements. In this case, I leave it you then,

Comment (dorukkilitcioglu) [reply .github/workflows/ci.yml:18]: I'm inclined to keep 3.6 until we get a report saying that 3.6 is causing issues.

File: textwiser/embeddings/word.py [hunk 5]
```diff
@@ -13,13 +13,15 @@
...
 from gensim.models import Word2Vec
+import numpy as np
 import re
 import torch
 import torch.nn as nn
 from transformers import AutoTokenizer, AutoModel, AutoConfig
 
 try:
-    from allennlp.modules.elmo import Elmo, batch_to_ids
+    import tensorflow as tf
```
Comment (kuppulur) [inline textwiser/embeddings/word.py:23]: Why do we need a try/except for import here? Is it because, we don't expect the user to install requirements_full?

**Review by kuppulur — COMMENTED** [review]
Overall LGTM!!! I have few minor comments and one deterministic error below.

**Minor comments**
If I install requirements following
Python 3.9.12; MacOS Monterey Version: 12.6.1
`pip install -r requirements.txt`
`pip install -r requirements_full.txt`

I get, 
`ERROR: pip's dependency resolver does not currently take into account all the packages that are installed. This behaviour is the source of the following dependency conflicts.
konoha 4.6.5 requires importlib-metadata<4.0.0,>=3.7.0, but you have importlib-metadata 5.2.0 which is incompatible.`

Running tests takes sometime, due to model downloads etc., not sure if this is a problem for us, but if it is, we can consider mocking most of these calls, to make tests run faster. 

**Test Errors**
I have 3 test errors from `tests.test_doc2vec.Doc2VecTest` all of them due to mismatch in tensors, after considering tolerance. One example below

**tw_model.transform(docs)**
    tensor([[-0.2276,  0.1449],
            [ 0.0791,  0.0202],
            [-0.1494,  0.2438]])

**expected.to(device)**
    tensor([[ 0.1922,  0.2392],
            [-0.0607,  0.0165],
            [ 0.0748, -0.0935]])


**tw_model(docs)**
    tensor([[-0.2276,  0.1449],
            [ 0.0791,  0.0202],
            [-0.1494,  0.2438]])

**expected.to(device)**
    tensor([[ 0.1922,  0.2392],
            [-0.0607,  0.0165],
            [ 0.0748, -0.0935]])

File: textwiser/embeddings/word.py (hunk 5 above)
Comment (dorukkilitcioglu) [reply textwiser/embeddings/word.py:23]: Correct, we don't expect the users to install `requirements_full`. Tensorflow is optional and you won't get it with `pip install textwiser`.

**Comment by dorukkilitcioglu** [comment]
> Overall LGTM!!! I have few minor comments and one deterministic error below.
> 
> **Minor comments** If I install requirements following Python 3.9.12; MacOS Monterey Version: 12.6.1 `pip install -r requirements.txt` `pip install -r requirements_full.txt`
> 
> I get, `ERROR: pip's dependency resolver does not currently take into account all the packages that are installed. This behaviour is the source of the following dependency conflicts. konoha 4.6.5 requires importlib-metadata<4.0.0,>=3.7.0, but you have importlib-metadata 5.2.0 which is incompatible.`
> 
> Running tests takes sometime, due to model downloads etc., not sure if this is a problem for us, but if it is, we can consider mocking most of these calls, to make tests run faster.
> 
> **Test Errors** I have 3 test errors from `tests.test_doc2vec.Doc2VecTest` all of them due to mismatch in tensors, after considering tolerance. One example below
> 
> **tw_model.transform(docs)** tensor([[-0.2276, 0.1449], [ 0.0791, 0.0202], [-0.1494, 0.2438]])
> 
> **expected.to(device)** tensor([[ 0.1922, 0.2392], [-0.0607, 0.0165], [ 0.0748, -0.0935]])
> 
> **tw_model(docs)** tensor([[-0.2276, 0.1449], [ 0.0791, 0.0202], [-0.1494, 0.2438]])
> 
> **expected.to(device)** tensor([[ 0.1922, 0.2392], [-0.0607, 0.0165], [ 0.0748, -0.0935]])

Re: doc2vec, you're probably missing the python hash seed: https://fidelity.github.io/textwiser/installation.html#test-your-setup

Re: pip dependency issues, was that on a new environment or an existing one? Also, it might be due to the two stage installation you had there. If you want to install everything including the optional recs, I'd suggest running `pip install -e .[full]` to install the library instead.

**Comment by kuppulur** [comment]
> > Overall LGTM!!! I have few minor comments and one deterministic error below.
> > **Minor comments** If I install requirements following Python 3.9.12; MacOS Monterey Version: 12.6.1 `pip install -r requirements.txt` `pip install -r requirements_full.txt`
> > I get, `ERROR: pip's dependency resolver does not currently take into account all the packages that are installed. This behaviour is the source of the following dependency conflicts. konoha 4.6.5 requires importlib-metadata<4.0.0,>=3.7.0, but you have importlib-metadata 5.2.0 which is incompatible.`
> > Running tests takes sometime, due to model downloads etc., not sure if this is a problem for us, but if it is, we can consider mocking most of these calls, to make tests run faster.
> > **Test Errors** I have 3 test errors from `tests.test_doc2vec.Doc2VecTest` all of them due to mismatch in tensors, after considering tolerance. One example below
> > **tw_model.transform(docs)** tensor([[-0.2276, 0.1449], [ 0.0791, 0.0202], [-0.1494, 0.2438]])
> > **expected.to(device)** tensor([[ 0.1922, 0.2392], [-0.0607, 0.0165], [ 0.0748, -0.0935]])
> > **tw_model(docs)** tensor([[-0.2276, 0.1449], [ 0.0791, 0.0202], [-0.1494, 0.2438]])
> > **expected.to(device)** tensor([[ 0.1922, 0.2392], [-0.0607, 0.0165], [ 0.0748, -0.0935]])
> 
> Re: doc2vec, you're probably missing the python hash seed: https://fidelity.github.io/textwiser/installation.html#test-your-setup
> 
> Re: pip dependency issues, was that on a new environment or an existing one? Also, it might be due to the two stage installation you had there. If you want to install everything including the optional recs, I'd suggest running `pip install -e .[full]` to install the library instead.

Re: Testing: Using PYTHONHASHSEED, all the tests pass.
Re: Installation: Yes, I did a two stage installation. Doing it `pip install -e ."[full]"` works.

**Review by kuppulur — APPROVED** [review]
LGTM! Installation works and all the tests pass.
